src/backend/optimizer/plan/createplan.c

```diff
@@ -4443,7 +4443,16 @@ create_ctescan_plan(PlannerInfo *root, Path *best_path,
 		if (!cteplaninfo->shared_plan)
 		{
 			RelOptInfo *sub_final_rel;
-			GangType	saved_gangType = root->curSlice->gangType;
+			GangType	saved_gangType = GANGTYPE_UNALLOCATED;
+			/*
+			 * Since topSlice is only initialized on the QD,
+			 * root->curSlice may be NULL in other roles. Check it first.
+			 */
+			Assert(Gp_role != GP_ROLE_DISPATCH || root->curSlice != NULL);
+			if (root->curSlice)
```

Comment:
why not
```suggestion
			if (Gp_role == GP_ROLE_DISPATCH && root->curSlice)
```
as below?